design/gateway-apis-implementation.md

```diff
@@ -1,204 +1,420 @@
-A conflict will **not** result in the whole HTTPRoute being rejected unless there are zero rules left.
+Contour will only ever update the status of HTTPRoute objects.
 
+Errors or conflicts here will render that rule invalid, but not the rest of the rules. Other valid rules will still be passed to Envoy.
+For each invalid rule, Contour will update status information with the rule and the reason. A conflict will not result in the whole HTTPRoute being rejected unless there are zero rules left.
 
 The output of this watcher is Envoy configuration.
 
 ### TLSRoute
-Contour will watch all TLSRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
-and will configure the associated routes.
-Configuration of TLSRoutes  will be subject to the rules around the `RouteGateways` field.
 
-Contour will only ever update the `status` of TLSRoute objects.
+Contour will watch all TLSRoute objects, and filter for entries matching a label selector in its configured Gateway's spec.listeners[].routes, and will configure the associated routes.
+Configuration of TLSRoutes will be subject to the rules around the RouteGateways field.
 
-Errors or conflicts here will render that section of the config invalid.
-Other valid sections will still be passed to Envoy.
+Contour will only ever update the status of TLSRoute objects.
+
+Errors or conflicts here will render that section of the config invalid. Other valid sections will still be passed to Envoy.
 
 The output of this watcher is Envoy configuration.
 
 ## Detailed Design
 
-### Configuration
-Contour will have an entry added to the config file for the Gateway it should watch.
+### Responsibilities
+
+#### Contour
+
+In the fully managed mode, Contour is responsible for managing the Envoy fleet which includes the Envoy Kubernetes Daemonset and the Envoy Kubernetes Service.
+
+#### Operator
+
+The operator is responsible for managing instances of Contour as well as any additional configuration items that are related to running Contour or Envoy.
+These pieces will be configured in the Contour Configuration CRD.
+
+### Configmap to CRD
+
+Currently, Contour gets its configuration from two different places, one is the configuration file represented as a Kubernetes configmap.
+The other are flags which are passed to Contour.
+
+This design will migrate the Contour configuration configmap to a full custom resource definition which will allow for better validation, easier processing, as well as represent a full Contour configuration including how Envoy is to be provisioned if the user desires.
+
+#### Benefits
+
+* Eliminates the need to translate from a CRD to a configmap (Like the Operator does today)
+* Allows for a place to surface information about configuration errors - the CRD status, in addition to the Contour log files
+* Allows the Operator workflow to match a non-operator workflow  (i.e. you start with a Contour configuration CRD)
```

Comment:
What exactly do you mean by this? Are you implying that [the Operator's Contour CRD](https://github.com/projectcontour/contour-operator/blob/main/api/v1alpha1/contour_types.go#L38-L50) goes away and is replaced by the new Contour config CRD? If so, I'm not clear where an operator user would configure things like the namespace to install Contour in, or the # of replicas in the Contour deployment, etc. But if that's not what you meant, disregard.

Reply:
Num of replicas, etc is a good point. Originally I had wanted to have a different card for the operator vs contour, but didn't get much support for the idea. 

We could do something like that but make the config crd a reference or something? Worth a discussion on. 

Or those fields are just ignored if not using the operator which feels bad as well.

Reply:
I think we're probably going to end up with two CRDs: the operator Contour/ContourDeployment CRD that describes how Contour should be deployed. This should reference a ContourConfig CRD, or reuse the same structs as the ContourConfig CRD, which the controller will watch and handle.

Reply:
Actually, now I've thought about it a little more, I think that using the ContourConfig Spec struct in the Operator's ContourDeployment object is the way to go, so you would end up with something like:
```yaml
kind: ContourDeployment
spec:
  deployment: <replicas and other Contour controller deployment settings>
  config:
    <include ContourConfigSpec struct here>
```

Doing this means that the Operator can simply take whatever is in the `config` stanza, and turn it into the Spec for a ContourConfig object, and that updating ContourConfig automatically updates the ContourDeployment's fields.

Reply:
Yeah, seems to me like we'll need to have 2 CRDs as well. Will think a little more about embedding the Spec struct vs. using a reference but we can definitely make one of those work.

Reply:
I'd prefer to not have the reference, having to manage 100 different objects is way too complex for what we're trying to do. That was the idea behind having HTTPProxy be able to work as a single resource, or many. 

To make it simpler, we could put all the CRDs in the Contour repo, and the operator could just reference those CRDs, so it's just a consumer of the `ContourDeployment` (or whatever name it gets).

Reply:
would be good to enumerate what the operator only information is like mentioned above (# of replicas etc.) and see what we need as fields and what we can achieve by convention (eg. namespace of created Contour could be the same the Config CRD is in etc.)

Reply:
Agreed on that last point, @sunjayBhatia, and agreed as well that putting all the CRDs in one place makes sense (probably the Contour repo, but at some point it may be worth extracting them out into a separate repo).

Reply:
The reason I suggested embedding also is because of things like Deployment and ReplicaSet already using that strategy.